### PR TITLE
@foxglove/velodyne-cloud

### PR DESCRIPTION
**Pull request by jhurliman**
This library transforms raw Velodyne packets into PointCloud classes that closely resemble the structure of PointCloud2 ROS messages. Decoding a full HDL32E scan in my Linux VM takes 5-6ms.

### CODE REVIEW

File: packages/velodyne/perf/transform.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (defunctzombie) [inline packages/velodyne/perf/transform.js:1]: Please convert this to typescript

File: packages/velodyne/src/Calibration.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    return Array(rows)
+      .fill(0)
+      .map((_row, x) =>
+        Array(cols)
+          .fill(0)
+          .map((_col, y) => fullFiring * block(x, y) + singleFiring * point(x, y) + offset),
+      );
+  }
+
+  static Data(model: Model): CalibrationData {
```
Comment (defunctzombie) [inline packages/velodyne/src/Calibration.ts:127]: `FromData` might be a more suitable name? Or maybe this belongs in a separate file on converters?

File: packages/velodyne/src/Calibration.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+import HDL32E_json from "./data/calibration/32db.json";
+import HDL64E_S21_json from "./data/calibration/64e_s2.1-sztaki.json";
+import HDL64E_S3_json from "./data/calibration/64e_s3-xiesc.json";
+import HDL64E_json from "./data/calibration/64e_utexas.json";
+import VLP16HiRes_json from "./data/calibration/VLP16_hires_db.json";
+import VLP16_json from "./data/calibration/VLP16db.json";
+import VLS128_json from "./data/calibration/VLS128.json";
+import VLP32C_json from "./data/calibration/VeloView-VLP-32C.json";
+
+export class Calibration {
```
Comment (defunctzombie) [inline packages/velodyne/src/Calibration.ts:15]: Class documentation please

File: packages/velodyne/src/RawPacket.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  distance(index: number): number {
+    return this.view.getUint16(4 + 3 * index, true);
+  }
+
+  intensity(index: number): number {
+    return this.data[4 + 3 * index + 2] as number;
+  }
+}
+
+export class RawPacket {
```
Comment (defunctzombie) [inline packages/velodyne/src/RawPacket.ts:36]: Split these two classes into files? Typically we try to keep 1 class per file - helps keep intent for each class clear.

File: packages/velodyne/src/RawPacket.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      const blockSize = RawPacket.BLOCK_SIZE;
+      const blockData = new Uint8Array(data.buffer, data.byteOffset + blockSize * i, blockSize);
+      this.blocks.push(new RawBlock(blockData));
+    }
+
+    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
+
+    this.gpsTimestamp = view.getUint32(1200, true);
+    this.factoryField1 = data[1204] as number;
+    this.factoryField2 = data[1205] as number;
```
Comment (jtbandes) [inline packages/velodyne/src/RawPacket.ts:67]: `as number` shouldn't be needed when indexing a Uint8Array?

File: packages/velodyne/src/Transformer.test.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+import { HDL32E_PACKET1 } from "../fixtures/packets";
+import { Calibration } from "./Calibration";
+import { PointCloud } from "./PointCloud";
+import { RawPacket } from "./RawPacket";
+import { Transformer } from "./Transformer";
+import { Model } from "./VelodyneTypes";
+
+function norm(x: number, y: number, z: number): number {
+  return Math.sqrt(x * x + y * y + z * z);
```
Comment (jtbandes) [inline packages/velodyne/src/Transformer.test.ts:15]:
```suggestion
  return Math.hypot(x, y, z);
```

**Comment by defunctzombie** [comment]
Please use a different name - possibly something with _ros_ in the name if appropriate or `velodyne-to-pointcloud`. My thinking is that if we layer need to convert to other formats this will have taken a top level _velodyne_ namespace.

File: packages/velodyne/perf/transform.js (hunk 1 above)
Comment (jhurliman) [reply packages/velodyne/perf/transform.js]: This is JS on purpose, you can't produce flame graphs for TS code.

File: packages/velodyne/src/RawPacket.ts (hunk 5 above)
Comment (jhurliman) [reply packages/velodyne/src/RawPacket.ts]: It is if you want a `number` and not a `number | undefined`.

File: packages/velodyne/perf/transform.js (hunk 1 above)
Comment (defunctzombie) [reply packages/velodyne/perf/transform.js]: What prevents us from producing flame graphs for ts code? We can run node with a transformer that transpiles the code on-the-fly (which is what ts-node does). How is the js different?

**Comment by amacneil** [comment]
👏 awesome

**Pull request merged**